Profiler API to request ReJIT with inliners

davmason opened the pull request:
This API is necessary for attaching profilers to be able to ReJIT methods and replace everything that uses the old IL.

I still have to do perf testing to make sure when no profiler is attached we don't regress anything (besides the small impact to overall working set from the tracking), and there are a few more scenarios I want to test but overall it is ready for review/feedback.

jkotas commented on src/vm/jitinterface.cpp:
```diff
@@ -7993,57 +7994,91 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
+        _ASSERTE(result == INLINE_PASS);
+        // We don't want to track the chain of methods, so intentionally use m_pMethodBeingCompiled
+        // to just track the methods that pCallee is eventually inlined in
+        pCallee->GetModule()->AddInlining(pOrigCaller, pCallee);
```
You should do this in `reportInliningDecision` instead. There is a bunch of checks that the JIT does after this, and so you will be getting a bunch of false positives by recording it here. `reportInliningDecision` is when the inlining decision is done and confirmed.

jkotas commented on src/vm/jitinterface.cpp:
```diff
@@ -7993,57 +7994,91 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
+        // Need to hold the ReJIT lock between when we check if a method should be inlined and
+        // when we report the inlining or there could be a race condition where a method is ReJITted after
+        // checking if it has default IL and before we report the inlining.
+        CrstHolder ch(&(ReJitManager::s_csGlobalRequest));
```
Taking a global lock here does not look good. We try to make the whole type loader / JIT process lock free.

noahfalk replied:
This caught my eye too in an earlier review, but I wasn't aware there was any official stance on it as long as overall performance testing had satisfactory results.

If I recall correctly I am probably guilty of taking the CodeVersionManager lock in Prestub path in order to determine which version of code should be invoked/jitted. The tiered compilation call-counter also has a lock around it. Not to say that doing a bad thing elsewhere should justify repeating it, but I at least had been undermining it and others perhaps are as well?

This particular case probably isn't that hard to eliminate the lock if we need to.

jkotas replied:
>  it as long as overall performance testing had satisfactory results.

It is not just about performance. It is also about correctness. For example, the correctness of this lock depends on a fragile assumption that the JIT asks for IL before calling `canInline`. It happens to be the case today, but there is nothing that guarantees it. The JIT/EE interface is meant to be order independent. The JIT is free to ask questions in any order it wants and still get correct answers.

noahfalk commented on src/vm/rejit.cpp:
```diff
@@ -685,12 +729,210 @@ HRESULT ReJitManager::UpdateActiveILVersions(
+                        {
+                            // This method has already been ReJITted, no need to request another ReJIT at this point.
+                            // The ReJITted method will be in the JIT inliner check below.
+                            continue;
```
Apologies if I have just paged this out but what happens when A inlines B by default and a profiler wants to do:
Edit function A
Edit function B
Revert the edit to A

If the profiler calls RejitWithInliner A, RejitWithInliner B, Revert A then it appears they are going to accidentally lose their edit to B that is now inlined in A again. Did we say they should use ReJIT instead of Revert and pass the original IL? 

It would probably help to write up the expected usage in a little design doc/documentation because it feels easy to accidentally do the wrong thing.

davmason replied:
I don't remember talking about this specific detail before. You are correct that the way they are currently implemented reverts will ignore any modified IL. 

I think the right course is documenting it and suggesting that the right behavior is either:
1. You only care about rejitting one method and revert will do what you want
2. You care about rejitting more than one method, and revert is dangerous, so RequestReJIT with the original IL

We could also change the current revert behavior from activating the original native code to effectively doing another rejit with the original IL, but I am not convinced it's worth it

noahfalk commented on src/vm/rejit.cpp:
```diff
@@ -718,16 +961,31 @@ HRESULT ReJitManager::BindILVersion(
+            // There could be a case where the method that a profiler requested ReJIT on also ends up in the 
+            // inlining graph from a different method. In that case we should override the previous setting,
+            // but we should never override a request to get the callback with a request to suppress it.
+            pILCodeVersion->SetSuppressReJITCallback(false);
```
Nit: Naming this SetEnableReJITCallback and flipping the meaning of the flag would avoid needing to reason about double negatives that occur here.

noahfalk commented on src/vm/rejit.cpp:
```diff
@@ -554,56 +630,24 @@ HRESULT ReJitManager::UpdateActiveILVersions(
-        _ASSERTE(pCodeVersionManager != NULL);
-        CodeActivationBatch * pCodeActivationBatch = mgrToCodeActivationBatch.Lookup(pCodeVersionManager);
-        if (pCodeActivationBatch == NULL)
+        hr = UpdateActiveILVersion(&mgrToCodeActivationBatch, pModule, rgMethodDefs[i], fIsRevert, static_cast<COR_PRF_REJIT_FLAGS>(0));
```
Rather than explicitly zeroing the flags, I suggest masking off the bits you don't want to flow. It is identical behavior now, but feels more intuitive about what I'd expect the behavior to be when a new flag gets added.

davmason replied:
I think this was actually a bug. I actually wanted to unconditionally set COR_PRF_REJIT_INLINING_CALLBACKS so the callback always was set for a method that had been passed to RequestReJIT, but instead was always clearing it.

I actually want to do `static_cast<COR_PRF_REJIT_FLAGS>(flags & COR_PRF_REJIT_INLINING_CALLBACKS)`

davmason commented:
What I just pushed was rebasing it on to the latest changes from master and fixing the conflicts with recent profiler changes. I haven't addressed code review feedback yet so don't waste time looking 😄

davmason commented:
Ok, all feedback should be accounted for now

jkotas commented on src/vm/jitinterface.cpp:
```diff
@@ -8008,56 +8009,72 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
 
 #ifdef PROFILING_SUPPORTED
-    if (CORProfilerPresent())
     {
```
Do we still need this extra scope?

davmason replied:
No, if was added for the lock holder and I forgot I was the one who added it when I removed the lock

davmason commented:
@davidwrighton can you take a look at the latest commit? I added a fake GetExposedObject for the default LoaderAllocator.

davidwrighton commented on src/vm/crossloaderallocatorhash.inl:
```diff
@@ -1099,7 +1099,23 @@ LAHASHDEPENDENTHASHTRACKERREF CrossLoaderAllocatorHash<TRAITS>::GetDependentTrac
+                {
+                    // Global LoaderAllocator does not have an exposed object, so create a fake one
+                    exposedObject = AllocateObject(MscorlibBinder::GetExistingClass(CLASS__OBJECT));
+                    m_globalDependentTrackerRootHandle = GetAppDomain()->CreateHandle(exposedObject);
```
You need to clean this handle up. See uses of RegisterHandleForCleanup where we create the other handles. Otherwise, you'll have created a memory leak for collectible assemblies.

davmason replied:
I added the RegisterHandleForCleanup and am going to check in since it's the only thing you mentioned. If you have further feedback let me know tomorrow and I'll address it in a new PR.
